src/common/src/com/microsoftopentechnologies/tooling/msservices/helpers/auth/AADManagerImpl.java

```diff
@@ -0,0 +1,637 @@
+/**
+ * Copyright 2014 Microsoft Open Technologies Inc.
+ * <p/>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p/>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p/>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.microsoftopentechnologies.tooling.msservices.helpers.auth;
+
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonSyntaxException;
+import com.google.gson.reflect.TypeToken;
+import com.microsoftopentechnologies.aad.adal4j.AuthenticationContext;
+import com.microsoftopentechnologies.aad.adal4j.AuthenticationResult;
+import com.microsoftopentechnologies.aad.adal4j.PromptValue;
+import com.microsoftopentechnologies.tooling.msservices.components.AppSettingsNames;
+import com.microsoftopentechnologies.tooling.msservices.components.DefaultLoader;
+import com.microsoftopentechnologies.tooling.msservices.components.PluginSettings;
+import com.microsoftopentechnologies.tooling.msservices.helpers.NotNull;
+import com.microsoftopentechnologies.tooling.msservices.helpers.StringHelper;
+import com.microsoftopentechnologies.tooling.msservices.helpers.azure.AzureCmdException;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class AADManagerImpl implements AADManager {
+    private static AADManager instance;
+    private static Gson gson;
+
+    private Map<UserInfo, Map<String, AuthenticationResult>> authResultByUserResource;
```

Comment:
As I understand it, the inner map object associates *resource* URLs with *access tokens* and the outer map associates *users* with the inner map. Could we put some documentation here to indicate that that's what this map is?